### ISSUE TITLE
Wrong result when calling 'multi-line' lambda inside @tf.function

### ISSUE THREAD

**Issue by andrewsali** [tensorflow/tensorflow#39832]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux, Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.2.0-rc4-8-g2b96f3662b 2.2.0
- Python version: 3.7.6
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

When decorating a function using `tf.function` that uses a lambda that extends to multiple lines,
only the first line is considered. When not decorating the function using `tf.function` the result is correct.

**Describe the expected behavior**

`tf.function` should not alter the function behaviour. In the example below, it prints `1` instead of `0`. Removing the `tf.function` decorator gives the correct result. Moving the lambda into a single line gives again a correct result.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

```
import tensorflow as tf

a = (
     lambda y: lambda x: x*y 
     - y
    )(1)

@tf.function
def test_lambda():
    tf.print(a(1))
    
test_lambda()
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

The issue was noticed after applying black formatting changed the code output.

**Comment by ravikyram** [comment]
I have tried in colab with TF version 2.2 , nightly version and was able to reproduce the issue.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/313709a5c0c9e445aee01e37d8c1d99f/untitled36.ipynb).Thanks!

**Comment by mdanatg** [comment]
This looks like an underlying Python bug:

```
l = (
     lambda y: lambda x: x*y
     - y
    )

import inspect
print(inspect.getsource(l))
```

Output:

```
     lambda y: lambda x: x*y
```

It looks like the parser misses the second line.

To fix this, we'll need to work around this bug.

**Comment by andrewsali** [comment]
Is there any update maybe on this issue? It is concerning that code formatting / linting cannot be safely applied with tf.function

**Comment by mdanatg** [comment]
We have confirmed this is caused by a bug in Python, but I think we'll have a workaround in TF 2.4. In the mean time, these practices should help avoid running into this bug:

1. Avoid placing lambdas inside expressions that could break them across multiple lines, and instead define them on a separate line:

```
partial = lambda y: lambda x: x*y - y
l = partial(1)
```

2. Prefer using normal local functions when the body might be too long to fit on a single line:

```
def partial(y):
  return lambda x, y: x*y - y
l = partial(1)
```

**Comment by mdanatg** [comment]
Quick update: a more robust fix for this is now in and should be available in tf-nightly and TF 2.4 later this year. The new parsing method should robustly process all cases, and raises an error when it can't determine the proper function with certainty (which should only happen in rare cases).

**Comment by andrewsali** [comment]
Fantastic, thanks!
